server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -79,132 +97,327 @@ public NodeConnectionsService(Settings settings, ThreadPool threadPool, Transpor
+                transportService.connectToNode(discoveryNode);
+                consecutiveFailureCount.set(0);
+                logger.debug("connected to {}", discoveryNode);
+                onCompletion(ActivityType.CONNECTING, null, disconnectActivity);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                assert Thread.holdsLock(mutex) == false : "mutex unexpectedly held";
+                final int currentFailureCount = consecutiveFailureCount.incrementAndGet();
+                // only warn every 6th failure
+                final Level level = currentFailureCount % 6 == 1 ? Level.WARN : Level.DEBUG;
+                logger.log(level, new ParameterizedMessage("failed to connect to {} (tried [{}] times)",
+                    discoveryNode, currentFailureCount), e);
+                onCompletion(ActivityType.CONNECTING, e, disconnectActivity);
+            }
+
+            @Override
+            public String toString() {
+                return "connect to " + discoveryNode;
+            }
+        });
+
+        private final Runnable disconnectActivity = new AbstractRunnable() {
+            @Override
+            protected void doRun() {
+                assert Thread.holdsLock(mutex) == false : "mutex unexpectedly held";
+                transportService.disconnectFromNode(discoveryNode);
+                consecutiveFailureCount.set(0);
+                logger.debug("disconnected from {}", discoveryNode);
+                onCompletion(ActivityType.DISCONNECTING, null, connectActivity);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                assert Thread.holdsLock(mutex) == false : "mutex unexpectedly held";
+                consecutiveFailureCount.incrementAndGet();
+                // we may not have disconnected, but will not retry, so this connection might have leaked
+                logger.warn(new ParameterizedMessage("failed to disconnect from {}, possible connection leak", discoveryNode), e);
+                assert false : "failed to disconnect from " + discoveryNode + ", possible connection leak\n" + e;
+                onCompletion(ActivityType.DISCONNECTING, e, connectActivity);
+            }
+        };
+
+        ConnectionTarget(DiscoveryNode discoveryNode) {
+            this.discoveryNode = discoveryNode;
+        }
+
+        Runnable connect(ActionListener<Void> listener) {
+            return addListenerAndStartActivity(listener, ActivityType.CONNECTING, connectActivity,
+                "disconnection cancelled by reconnection");
+        }
+
+        Runnable disconnect(ActionListener<Void> listener) {
+            return addListenerAndStartActivity(listener, ActivityType.DISCONNECTING, disconnectActivity,
+                "connection cancelled by disconnection");
+        }
+
+        Runnable ensureConnected(ActionListener<Void> listener) {
```

Comment:
Is it possible that ensureConnected and connect/disconnect are called in different threads at the same time? I'm not sure how we're protecting listeners from races

Reply:
Yes, they can be called in different threads, but we only ever read or write `listeners` under the `mutex`. The listeners are never called under the `mutex` so it is possible that the notifications happen out of order, but this is benign.